### ISSUE TITLE
Development Pods Test target doesn't see private headers

### ISSUE THREAD

**Issue by GeroHerkenrath** [CocoaPods/CocoaPods#5384]
- [x] I've read and understood the [*CONTRIBUTING guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).
# Report
## What did you do?

Created a new pod, added two classes, one private and one public, plus tests for both.
## What did you expected to happen?

I expected both test classes to succeed and see all necessary headers (of the private and public class)
## What happened instead?

The tests for the private class couldn't run because they didn't find the header file.
### Stack

```
   CocoaPods : 1.0.0
        Ruby : ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin15]
    RubyGems : 2.6.3
        Host : Mac OS X 10.11.4 (15E65)
       Xcode : 7.3.1 (7D1014)
         Git : git version 2.7.4 (Apple Git-66)
Ruby lib dir : /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib
Repositories : master - https://github.com/CocoaPods/Specs.git @ 6e0c5b27c2af7074b4c4776610d24251aa63780a
```
### Installation Source

```
Executable Path: /usr/local/bin/pod
```
### Plugins

```
cocoapods-deintegrate : 1.0.0
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.0.0
cocoapods-try         : 1.0.0
```
## Project that demonstrates the issue

https://github.com/GeroHerkenrath/DevPodProblem
## Workaround

As noted in the readme of my demo project the issue isn't a biggie for public classes. Before Cocoapods 1.0 I was able to just use `#import "myHeader.h"`, now I have to use `#import <MyLib/MyHeader.h>` which isn't a big deal. But for the private headers this doesn't work, since they're not visible at all. Yet, they obviously need testing, hence I manually added `"${PODS_ROOT}/../../<podname>/Classes/Private"` to the test target's search paths. 

I think several people having filed issues might in the end have run into the same problem, too. I've noted that the [issue](https://github.com/CocoaPods/CocoaPods/issues/5341) with the derived data makes it hard to figure out what's wrong when you're merely upgrading existing pods to 1.0, but I'm not sure. Playing around with the `inherit! :search_paths` seemed to work at first, but that was due to the derived data not being properly cleared (once I did this the problem was back again). 

So there you have it, hopefully this helps some others.

(I should add that I\m relatively new to cocoapods, so if I made any wrong assumptions in the first place I apologize)

**Comment by neonichu** [comment]
I believe this is working as it should, private headers should not be exposed by dependencies. AFAIK, this was never intended to work differently, but there might have been bugs which inadvertently exposed private headers.

**Comment by GeroHerkenrath** [comment]
This is not about regular projects seeing them, but the development project. The unit tests, more specifically. I don't think it's that in the past that was a mistake, since in the past the private headers were visible to my unit tests, but not to the example project itself.
I'm specifically talking about the created template workspace that results from `pod lib create`. Which is meant to provide you with an Example project (if you specify so) and a means to test your new library. Having private headers is a common thing and they should be unit tested, obviously.

**Comment by benrudhart** [comment]
I'd also love to get this to work for Objective-C! Thank you @GeroHerkenrath for posting this issue here.

One "workaround" might be to use Swift instead of Objective-C. Import the framework with `@testable import <Frameworkname>` in your testfile and you'll be able to access the classes/ methods with an access level set to `internal`

**Comment by GeroHerkenrath** [comment]
Glad I could help someone (and that I'm not the only one facing this in the first place). It makes sense that it works with swift, since, after all, there are no header files. Either it finds a class (and thus its entire implementation) or not.

The workaround I proposed works, I have released several minor versions of the pod I'm working on (for work) with the header search paths defined manually. Since it's only done in the Example (i.e. development) project and not the pod settings it's robust against updating the `Podfile` and `.specs` files (and running `pod update` and `pod install`). Neither does it interfere with the pod itself (i.e. somebody using my pod normally is not affected by this at all).

I just think the project template created by `pod lib create` should be updated so that it already has the search paths to potential private classes defined. Or at least that should be documented somewhere (in the cocoapods guides, where `pod lib create` is explained or where the `public_header_files` flag for a spec file is explained).

**Comment by benrudhart** [comment]
Thank you for the update and also for the example repo. It does help but there is one minor thing I'd like to point out:
If there is only one test file importing a private header directly (as you do in `MyPrivateClassSpec.m` via `#import "MyPrivateClass.h"`) without any import of a public class the compiler will throw an error -> you have to use
`@import DevPodProblem;` 
or `#import <DevPodProblem/AnyPublicClassOfTheFramework.h>`
 before the import of the private class in order to satisfy the compiler.

This isn't necessary in your project bc. the public class is imported in the test target in another file (`MyPublicClassSpec.m`). 

This might also help others bc. I reproduced your steps in one of my projects and it didn't work bc. I forgot to import the module/ a public class of this module.

**Comment by GeroHerkenrath** [comment]
Oh, that's a keen observation. I didn't notice that yet, but it makes sense. After all, the implementation of the private classes is still part of the module (i.e. its binary). Including *only* the private headers must lead to an error. I hadn't thought about that, since in my projects so far I have always also had public classes and usually was writing the tests for those first, so I had included public headers or the module. I will eventually comment that in my example so it"s clear. I can't come up with a way to avoid `@import`ing the module or a public class on the top of my head, but I think for a test target that's fine anyways.

**Comment by benmarten** [comment]
I was running into the same issue and fixing it the way your proposed didn't work for me somehow, I couldn't figure out why not.
Anyways, I ended up seting up a private spec repo, where i can push my private development pods to, that way they are actually included correctly and there is no mess with the header imports.

<img width="259" alt="screen shot 2016-08-12 at 4 09 57 pm" src="https://cloud.githubusercontent.com/assets/5960238/17639862/a98191a4-60ab-11e6-91e8-f84189e54204.png">

**Comment by GeroHerkenrath** [comment]
Doesn't that include a dependency to your private repo for anyone who uses your final pod?

And yes, in this constellation it is impossible to work concurrently on the two libs, but if you group your classes and functionality properly that shoudn't be a biggie.

In general however, I think you were facing a slightly different problem, weren't you? Note that the problem I described is not about including an entire library (or framework for that matter) and writing tests for that (which strikes me as a bad idea in the first place), it's just about "grouping" classes of my one library (or pod) into private and public ones. The goal being better readability for anyone using the pod. 

Of course that's more of a cosmetic thing. Since the pod is open source anyone can look at the private classes in the end (the way cocoapods downloads and includes pods into your workspace doesn't respect private and public sub-folders anyways, at least not per default I think. All classes/files of a pod are dumped into one folder). Structuring my development project accordingly makes it more accessible, though, at least in my humble opinion. And for that I then need to be able to write tests for the private classes, hence I need to somehow tell the compiler where their headers are.

If I put all private classes into a separate pod I circumvent the issue, at least as long as I keep them public classes of that pod, but that requires to divide the whole class structure much more strongly than a private/public division. Two libraries are, I'd say, more distinct from each other than the private and public classes of a single library. Plus, you get the "overhead" of another dependency.

If the strategy worked for your project, great. I definitely appreciate the feedback (and am now curious why including the headers the way I described didn't work for you, btw). I think your solution is, however, one for a slightly different issue.
